android-components/components/lib/crash/build.gradle

```diff
@@ -93,3 +96,59 @@ apply plugin: "org.mozilla.telemetry.glean-gradle-plugin"
 apply from: '../../../android-lint.gradle'
 apply from: '../../../publish.gradle'
 ext.configurePublish(config.componentsGroupId, archivesBaseName, project.ext.description)
+
+/*
+ * Remote crash metrics logic.
+ *
+ * The below pulls metrics and pings from gecko-dev and ensures that they are processed by the glean
+ * parser.
+ */
+
+// The directory to store downloaded files.
+def gleanCrashFilesDir = new File(buildDir, "glean_crash")
+
+task downloadCrashMetrics(type: DownloadCrashGleanDefinitions) {
+    tag = "15785ab3b7d9cf103ad2ffc032e22c9fd3dc30b0"
+    target = gleanCrashFilesDir
+}
+
+// These task names are defined in the glean gradle plugin
+android.libraryVariants.all { variant ->
+    tasks.getByName("gleanGenerateMetricsSourceFor${variant.name.capitalize()}").dependsOn downloadCrashMetrics
+    tasks.getByName("gleanGenerateMetricsDocsFor${variant.name.capitalize()}").dependsOn downloadCrashMetrics
+}
+
+// Override the files that the glean gradle plugin will use (to add those we download)
+ext.gleanYamlFiles = [
+        "${projectDir}/metrics.yaml",
+        "${gleanCrashFilesDir}/metrics.yaml",
+        "${gleanCrashFilesDir}/pings.yaml",
+        "${gleanCrashFilesDir}/tags.yaml"
+]
+
+import java.nio.file.Paths;
+
+class DownloadCrashGleanDefinitions extends DefaultTask {
```

Comment:
An option here is to add `compileOnly project(":browser-engine-gecko")` to https://github.com/mozilla-mobile/firefox-android/blob/5a961111addca08dcaa402d2a79b05920c04e731/android-components/components/lib/crash/build.gradle#L77. The comment actually claims it already does, which is outdated :).

Then it should be possible to get to this yaml file through the geckoview.aar? There would ideally be a separate step where we refactor this into a separate `lib-crash-glean` library, similar to `lib-crash-sentry`, but that's out of scope and can definitely be done later.

Reply:
Thinking about it more, I think we actually have to do something like the above here, because this currently fetches these files for a specific hard-coded Github tag, which can easily become outdated/incorrect and require manual updates on different branches. With the above, we would automatically use the right version of these .yaml files that match the Gecko version being used by the app (no matter what branch it is build from.)

Reply:
Another thought: If we're just going to fetch these files from a specific tag, and decide that's OK, we could also copy this files into the tree. We can keep them up-to-date with a Github action too until we move Fenix to M-C. This way we don't need to download them, as part of the build.

Reply:
> The comment actually claims it already does, which is outdated :).
Haha, I was wondering about that, thanks for validating my concern.

As far as the tag/version of the files being outdated: my reasoning was that it shouldn't actually _matter_ if they are outdated, because changes should only be additive in nature (as far as metrics are concerned) and without adding the new code to fill in new metrics, it wouldn't make a difference. But from a development perspective, it _would_ be nice if it always fetched the correct related version for a particular release. I'm not sure whether these files are actually included in the `geckoview.aar` at all though?

---

> Another thought: If we're just going to fetch these files from a specific tag, and decide that's OK, we could also copy these files into the tree. We can keep them up-to-date with a Github action too until we move Fenix to M-C. This way we don't need to download them, as part of the build.

We did consider just having the files in-tree (that would have made this simpler, obviously) but we were hesitant to have them diverge (though ultimately it would just be another update/programming task, and divergence in this way isn't much different than having the fixed tag).

I'm okay with having them in-tree. That seems like it would address both the offline-build issue (sorry about that, I didn't actually realize that offline builds were supported) and `probe-scraper` integration (the current approach requires a manual project be added). I guess `probe-scraper` would just have duplicates (under different project names) of the metrics/pings?

Reply:
If they were in-tree, the only potential issue is that one platform (desktop, android, etc) is updated but other platforms are forgotten to be updated. But I think that's more-or-less a non-issue, since development to fill out new metrics will be needed anyway for each platform (and _maybe_ there will be platform-specific metrics).

We wanted the metrics to be the same for purposes of querying and joining ping data easily, however I think that's probably not that much of a concern with separate metrics definitions if we just make sure to use the same metrics names when applicable? I'm speaking somewhat naively here though, as I've not personally queried Glean ping data, and I don't know whether e.g. it's easy to join queries across project/application labels (or if that doesn't matter whatsoever).

Reply:
>  I'm not sure whether these files are actually included in the geckoview.aar at all though?

They are not currently, but we could technically add them. However, I think we're both leaning towards just adding the .yaml files to the tree now to prevent the issues above. The plan is for this repo to migrate to M-C later this year, at which point we can remove the duplication, and in the meantime the duplication isn't any worse than fetching a specific git tag. In both cases, a change in this repo needs to land to keep the files in sync with any changes happening on M-C. We either have to change the git tag or update the files directly. So that difference doesn't matter, it seems, plus then we can still build while offline, and avoid intermittent CI issues too.

If the files in M-C change a lot, we can hook up a Github action to open a PR and keep the files in sync. We have a few of those already.

Does that sound OK? @afranchuk @badboy @travis79 

@afranchuk Two minor points about landing this: 
- I think we can land these changes in one PR when ready. 
- We have an early code freeze (next week) for this cycle, because we're moving Fenix into this repo, as part of the Monorepo project and preparation. So, we should land this change on Feb. 14, first day of the 112 cycle. This also makes sure that the new metrics start at the beginning of 112, which is a nice consequence.

Reply:
@afranchuk The build is actually [failing](https://github.com/mozilla-mobile/firefox-android/pull/671/checks?check_run_id=11045687489) now, which is unexpected, given that the change to fetch .yaml files is part of this PR and Glean code generation should've happened?

```
[task 2023-02-01T21:18:10.581Z] e: /builds/worker/checkouts/vcs/android-components/components/lib/crash/src/main/java/mozilla/components/lib/crash/service/GleanCrashReporterService.kt: (20, 50): Unresolved reference: Pings
[task 2023-02-01T21:18:10.582Z] e: /builds/worker/checkouts/vcs/android-components/components/lib/crash/src/main/java/mozilla/components/lib/crash/service/GleanCrashReporterService.kt: (27, 50): Unresolved reference: Crash
[task 2023-02-01T21:18:10.583Z] e: /builds/worker/checkouts/vcs/android-components/components/lib/crash/src/main/java/mozilla/components/lib/crash/service/GleanCrashReporterService.kt: (91, 25): Unresolved reference: Pings
...
```

Reply:
Hmm, that is unusual. I haven't seen any failures with what I've written (from various states of change wrt build state). Changing to local files will likely smooth that out.

Reply:
Weirdest thing: I just changed my local copy to have copied definitions and removed all the downloading stuff, and now I'm seeing those same unresolved reference errors.

Reply:
Nevermind, my rebase decided to stack rather than replace a commit (so it still had downloaded files). With the local files it seems okay. So I guess there was some way to get it into a state where it wouldn't generate the downloaded files. I would have thought the task dependencies were set up to avoid that :shrug:

Reply:
Ok, nice! The PR is green now too, and I like how this simplifies the build. From my perspective, this is all good now, but would like @badboy and @travis79 to check our conversation and conclusion here too :).

@afranchuk do you still want to keep that other [PR](https://github.com/mozilla-mobile/firefox-android/pull/686) separate? I think including it all here is fine.

Reply:
After reading the conversation here, if keeping a separate copy of the files is no more or less of a pain than including them in the geckoview aar file and pulling them out, and makes offline building easier, I'd say that is fine. Since these metrics are organized by the application they are sent from, there shouldn't be any problem having the same names for metrics in different files/apps as long as the probe-scraper is configured correctly.

Reply:
@csadilek I can remove the other PR. I only added it because I was unsure of what process should be followed WRT traceability and linking in bugzilla.

Reply:
@afranchuk OK, thanks, I will close the other PR and link it to this one then.